utils/src/main/java/com/airbnb/reair/common/DistCpWrapperOptions.java

```diff
@@ -164,21 +168,46 @@ public long getDistCpPollInterval() {
   }
 
   /**
-   * Returns the timeout that should be used, given a filesize.
-   * Helps determine whether to use dynamic timeout or not, and handles logic for that.
-   * @param filesizeBytes filesize of files to be copied in bytes
-   * @return The timeout to be used, in millis
+   * Returns the distcp timeout in milliseconds according to options set.
+   * @param fileSizes File sizes of the files to copy.
+   * @param maxConcurrency The number of mappers in distcp.
+   * @return The timeout in milliseconds for distcp.
    */
-  public long getDistcpTimeout(long filesizeBytes) {
+  public long getDistcpTimeout(List<Long> fileSizes, long maxConcurrency) {
     if (distcpDynamicJobTimeoutEnabled) {
-      long timeout = ((long) Math.ceil(filesizeBytes / 1e9)) * distcpDynamicJobTimeoutMsPerGb;
-      long minTimeout = distcpDynamicJobTimeoutMin;
+      long bytesPerMapper = computeLongestMapper(fileSizes, maxConcurrency);
+      long baseTimeout = distcpDynamicJobTimeoutBase;
       long maxTimeout = distcpDynamicJobTimeoutMax;
-      timeout = Math.max(minTimeout, timeout);
-      timeout = Math.min(maxTimeout, timeout);
+      long msPerGb = distcpDynamicJobTimeoutMsPerGbPerMapper;
+      long adjustment = ((long) Math.ceil(bytesPerMapper / 1e9) * msPerGb);
+      long timeout = Math.min(maxTimeout, baseTimeout + adjustment);
```

Comment:
What's the reason for baseTimemout + adjustment vs. just adjustment?

Reply:
Scheduling delay.

Reply:
Plus overhead

Reply:
Sounds good. Maybe add a comment about it?

Reply:
see above